lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp

```diff
@@ -265,23 +262,97 @@ LogicalResult torchAlphaToMhloTensor(ConversionPatternRewriter &rewriter,
   return success();
 }
 
+// Return a tensor filled with the given value of the given elementType, of
+// which shape equals the given tensor. Dynamic shape is supported, too.
+// This function do not check the compatiblity of storage type T and
+// elementType!
+template <typename T>
+Value getConstTensorLike(PatternRewriter &rewriter, Operation *op, T val,
+                         Value tensor, Type elementType) {
+  Location loc = op->getLoc();
+  // Generate zero rank const tensor.
+  auto tensorType = tensor.getType().cast<RankedTensorType>();
+  auto constType = RankedTensorType::get({}, elementType);
+  auto constAttr = DenseElementsAttr::get(constType, /*ArrayRef<T>*/ {val});
+  auto constOp = rewriter.create<mhlo::ConstantOp>(loc, constType, constAttr);
+
+  // Broadcast it
+  auto outType = tensorType.cloneWith(tensorType.getShape(), elementType);
+  if (tensorType.hasStaticShape()) {
+    return broadcastUnaryOperandStatic(rewriter, op, constOp, outType);
+  } else {
+    auto tensorShape = rewriter.create<shape::ShapeOfOp>(loc, tensor);
+    return rewriter.create<mhlo::DynamicBroadcastInDimOp>(
+        loc, outType, constOp, tensorShape, rewriter.getI64TensorAttr({}));
+  }
+}
+// template initialization
+template Value getConstTensorLike<float>(PatternRewriter &, Operation *, float,
+                                         Value, Type);
+template Value getConstTensorLike<int32_t>(PatternRewriter &, Operation *,
+                                           int32_t, Value, Type);
+
+// Broadcast the input tensor of 0 rank to a tensor with the same shape as the
+// given tensor. Supporting dynamic shape.
+Value broadcastZeroRankTensorToTensorLike(PatternRewriter &rewriter,
+                                          Operation *op, Value input,
+                                          Value tensor, Type elementType) {
+  Location loc = op->getLoc();
+  auto inputType = input.getType().cast<RankedTensorType>();
+  auto tensorType = tensor.getType().cast<RankedTensorType>();
+  if (inputType.getRank() != 0) {
+    op->emitError("Only zero ranked tensor is supported");
```

Comment:
Either rewriter.notifyMatchFailure if this is a limitation of the pattern or the check should be in the verifier of the op.
Could the return type be FailureOr<...> ? cc @jpienaar

Reply:
Good suggestion, we will change the codes.